conn.go

```diff
@@ -3,32 +3,46 @@
 package preconn
 
 import (
+	"bytes"
+	"errors"
+	"io"
 	"net"
 )
 
-// Conn is a net.Conn that supports replaying.
+// Conn is a net.Conn that supports replaying. Reads are not concurrency-safe.
 type Conn struct {
 	net.Conn
-	head []byte
+	head         io.Reader
+	consumedHead bool
 }
 
 // Wrap wraps the supplied conn and inserting the given bytes at the head of the
 // stream.
 func Wrap(conn net.Conn, head []byte) *Conn {
-	return &Conn{
-		Conn: conn,
-		head: head,
-	}
+	return WrapReader(conn, bytes.NewReader(head))
+}
+
+// WrapReader wraps the supplied conn, reading from 'head' first.
+func WrapReader(conn net.Conn, head io.Reader) *Conn {
+	return &Conn{conn, head, false}
 }
 
 // Read implements the method from net.Conn and first consumes the head before
 // using the underlying connection.
-func (conn *Conn) Read(b []byte) (int, error) {
-	n := copy(b, conn.head)
-	conn.head = conn.head[n:]
-	if n > 0 {
-		return n, nil
+func (conn *Conn) Read(b []byte) (n int, err error) {
+	if conn.consumedHead {
+		return conn.Conn.Read(b)
+	}
+	// "Read conventionally returns what is available instead of waiting for more."
+	// - https://golang.org/pkg/io/#Reader
+	// Thus if we read off conn.head, we return immediately rather than continuing to conn.Conn.
```

Comment:
Technically, stuff from conn.Read may actually be available here, we just don't know because we haven't tried :)

Reply:
Yes.  What would you do?  Start a read in another routine?  My thinking is that the reader can just call `Read` again if they need to.

FWIW, this has been the behavior since https://github.com/getlantern/preconn/pull/2.  Effectively all I did here is add a comment.

Reply:
Oh, I see what you mean. #2 dealt with the special case where something is available from the head and nothing is available from the wrapped Conn. Maybe you can clarify just that in the comment, something like:

`Per convention, read will not wait to fill the supplied buffer. If we've read anything off conn.head, don't bother waiting to read from conn.Conn since that may have nothing available and will cause Read() to block even though some data is already available. This may cause us to miss data that was already available on conn.Conn but avoids deadlocks.`.

Reply:
Sure

Reply:
https://github.com/getlantern/preconn/pull/3/commits/c4de1ce100bb1fe8137484d7fa9c4521f8e62c70